packages/wallet/src/wallet.ts

```diff
@@ -263,48 +326,63 @@ export class DemexWallet {
 
   public async estimateTxFee(
     messages: readonly EncodeObject[],
-    feeDenom: string,
-  ) {
-    const denomGasPrice = await this.getGasPrice(feeDenom);
+    denom: string,
+    granter?: string,
+  ): Promise<StdFee> {
+    const denomGasPrice = await this.getGasPrice(denom);
+
+    const totalGasCost = await this.getTotalGasCost(messages);
 
-    let totalGasCost = BN_ZERO;
-    for (const msg of messages) {
-      const gasCost = await this.getGasCost(msg.typeUrl);
-      totalGasCost = totalGasCost.plus(gasCost);
-    }
     let totalFees = totalGasCost.times(denomGasPrice ?? BN_ZERO);
 
     // override zero gas cost tx with some gas for tx execution
     // set overall fee to zero, implying 0 gas price.
     if (totalGasCost.isZero()) {
-      totalGasCost = TxDefaultGasCost;
       totalFees = BN_ZERO;
     }
 
     return {
-      amount: [{
-        amount: totalFees.toString(10),
-        denom: feeDenom,
-      }],
-      gas: totalGasCost.toString(10),
+      amount: [
+        {
+          amount: totalFees.toString(10),
+          denom,
+        },
+      ],
+      gas: DefaultGas.toString(10),
+      granter,
     };
   }
 
-  public async reloadAccountSequence() {
-    if (this.sequenceInvalidated) this.sequenceInvalidated = false;
+  private async getTotalGasCost(messages: readonly EncodeObject[]) {
+    let totalGasCost = BN_ZERO;
+    for (const message of messages) {
+      const gasCost = await this.getGasCost(message.typeUrl);
+      totalGasCost = totalGasCost
+        .plus(gasCost ?? BN_ZERO)
+        .plus(await this.addAdditionalGasCost(message));
+    }
+    return totalGasCost;
+  }
 
-    const info = await this.reloadAccountInfo()
-    const pubkey = this.accountInfo?.pubkey ?? {
-      type: "tendermint/PubKeySecp256k1",
-      value: this.publicKey.toString("base64"),
-    };
-    if (info) {
-      this.accountInfo = {
-        ...info,
-        pubkey,
-      };
+  private async addAdditionalGasCost(message: EncodeObject) {
+    switch (message.typeUrl) {
+      case TxTypes.MsgExec: return await this.getExecGasCost(message);
     }
+    return BN_ZERO;
+  }
+
+  private async getExecGasCost(message: EncodeObject) {
+    const { msgs } = message.value as MsgExec;
+    return await this.getTotalGasCost(msgs);
   }
+
+  public async reloadAccount(reloadAddresses: ReloadAddresses) {
+    const info = await this.reloadAccountInfo(reloadAddresses);
+    if (!info) return;
+    const { pubkey, ...rest } = info;
+    this.accountStates[reloadAddresses.address] = { ...rest, sequenceInvalidated: false };
+  }
```

Comment:
recommend passing in the key used to index `this.accountStates` instead of custom `ReloadAddresses` type.